From 59d2b98d06a58cc9c37f5aba13fb2f01df5da792 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:08:01 -0400
Subject: [powerpc] introduce macro LOAD_REG_ADDR_PIC

Message-id: <1410545655-205645-253-git-send-email-dzickus@redhat.com>
Patchwork-id: 94281
O-Subject: [RHEL7 PATCH 252/626] powerpc: introduce macro LOAD_REG_ADDR_PIC
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 1c49abec677c7ff495837dbaafad8e12f09c29fc
Author: Kevin Hao <haokexin@gmail.com>
Date:   Tue Dec 24 15:12:05 2013 +0800

    powerpc: introduce macro LOAD_REG_ADDR_PIC

    This is used to get the address of a variable when the kernel is not
    running at the linked or relocated address.

    Signed-off-by: Kevin Hao <haokexin@gmail.com>
    Signed-off-by: Scott Wood <scottwood@freescale.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/ppc_asm.h b/arch/powerpc/include/asm/ppc_asm.h
index 971ba86..ec51f77 100644
--- a/arch/powerpc/include/asm/ppc_asm.h
+++ b/arch/powerpc/include/asm/ppc_asm.h
@@ -296,6 +296,11 @@ n:
  *   you want to access various offsets within it).  On ppc32 this is
  *   identical to LOAD_REG_IMMEDIATE.
  *
+ * LOAD_REG_ADDR_PIC(rn, name)
+ *   Loads the address of label 'name' into register 'run'. Use this when
+ *   the kernel doesn't run at the linked or relocated address. Please
+ *   note that this macro will clobber the lr register.
+ *
  * LOAD_REG_ADDRBASE(rn, name)
  * ADDROFF(name)
  *   LOAD_REG_ADDRBASE loads part of the address of label 'name' into
@@ -306,6 +311,14 @@ n:
  *      LOAD_REG_ADDRBASE(rX, name)
  *      ld rY,ADDROFF(name)(rX)
  */
+
+/* Be careful, this will clobber the lr register. */
+#define LOAD_REG_ADDR_PIC(reg, name)  \
+ bl 0f;    \
+0: mflr reg;    \
+ addis reg,reg,(name - 0b)@ha;  \
+ addi reg,reg,(name - 0b)@l;
+
 #ifdef __powerpc64__
 #define LOAD_REG_IMMEDIATE(reg,expr)  \
  lis     reg,(expr)@highest;  \
-- 
1.7.1